pkg/tenant/topics.go

```diff
@@ -257,30 +262,137 @@ func (s *Server) TopicUpdate(c *gin.Context) {
+		return
+	}
+
+	// User claims are required to verify that the user owns the topic
+	if claims, err = middleware.GetClaims(c); err != nil {
+		log.Error().Err(err).Msg("could not fetch claims from context")
+		c.JSON(http.StatusUnauthorized, api.ErrorResponse("could not fetch claims from context"))
+		return
+	}
+
 	// Get the topic ID from the URL and return a 400 response
-	// if the topic does not exist.
+	// if the ID is not parseable
 	var topicID ulid.ULID
 	if topicID, err = ulid.Parse(c.Param("topicID")); err != nil {
-		log.Error().Err(err).Msg("could not parse topic ulid")
-		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not parse topic ulid"))
+		log.Warn().Err(err).Msg("could not parse topic id")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not parse topic id"))
+		return
+	}
+
+	// Parse the request body for the confirmation token
+	confirm := &api.Confirmation{}
+	if err = c.BindJSON(confirm); err != nil {
+		log.Warn().Err(err).Msg("could not bind topic delete request")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not bind user request"))
+		return
+	}
+
+	// Sanity check that the ID in the request body matches the ID in the URL
+	if confirm.ID != topicID.String() {
+		log.Warn().Err(err).Msg("topic id in request body does not match topic id in URL")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("id in request body does not match id in URL"))
+		return
+	}
+
+	// Fetch the topic metadata from the database
+	var topic *db.Topic
+	if topic, err = db.RetrieveTopic(ctx, topicID); err != nil {
+		log.Error().Err(err).Str("topicID", topicID.String()).Msg("could not retrieve topic")
+		c.JSON(http.StatusNotFound, api.ErrorResponse("topic not found"))
+		return
+	}
+
+	// Verify that the user owns the topic
+	if claims.OrgID != topic.OrgID.String() {
+		log.Warn().Err(err).Str("user_org", claims.OrgID).Str("topic_org", topic.OrgID.String()).Msg("topic OrgID does not match user OrgID")
+		c.JSON(http.StatusForbidden, api.ErrorResponse("user is not authorized to delete this topic"))
+		return
+	}
+
+	// Send confirmation token if not provided
+	if confirm.ConfirmToken == "" {
+		// Create a short-lived confirmation token in the database
+		topic.ConfirmDeleteToken = secrets.CreateToken(16)
+		topic.ConfirmDeleteExpire = time.Now().Add(5 * time.Minute)
+		if err = db.UpdateTopic(ctx, topic); err != nil {
+			log.Error().Err(err).Str("topicID", topicID.String()).Msg("could not save topic")
+			c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not generate confirmation token"))
+			return
+		}
+
+		confirm.Name = topic.Name
+		confirm.ConfirmToken = topic.ConfirmDeleteToken
+		c.JSON(http.StatusOK, confirm)
+		return
+	}
+
+	// Verify that the confirmation token is correct
+	if confirm.ConfirmToken != topic.ConfirmDeleteToken {
+		log.Warn().Err(err).Msg("confirmation token does not match")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("unrecognized confirmation token"))
```

Comment:
I think we should return a 412 Precondition Failed here to let the user know they should try the delete without the confirmation token again and then use a valid confirmation token.

```suggestion
		c.JSON(http.StatusPreconditionFailed, api.ErrorResponse("unrecognized confirmation token"))
```